streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
(1) We should enable assertions for tests!!!
(2) I think there are some important difference between an assertion and an exception. 
  - assertions "model" invariants that are true if the code is correct (they are there to help devs to read the code -- similar to `final` -- if a method is called and I see an `assert param != null` I know what's going on)
 - for this reason, we should assert literally everywhere (ie, each method call should check all provided parameters, nulls, value ranges -> negative ok?, etc)
 - this keeps the code at a very narrow path and allows to limit the possibility of bug (as tests would detect those issues)
 - assertion should be use for all "internal" checks and should never hit if there is no bug
 - and because if assertions are use extensively, you want to remove them in production and thus disable them

On the other hand, exceptions are used to check stuff that is not in the control of the code. I want to create file, but don't have permission (or folder does not exist). That's nothing you can check with an assertion, but need to report in production via an exception. Also for a library like Stream, all public API parameters are check via exceptions, not assertions -- it's not in control of the library how methods are called.